rfcs/text/0003-web-rpc-hooks.md

```diff
@@ -0,0 +1,280 @@
+        }
+      }
+    },
+    failure: (state, action) => {
+      return {
+        ...state,
+        user: {
+          ...state.user,
+          loading: false,
+          error: action.payload,
+        }
+      }
+    },
+  }
+});
+
+function Component() {
+  const user = useUser(args);
+  return <div>Hello {user.name}</div>
+}
+```
+
+The major drawback to this pattern is it is less familiar to people, and can cause confusion as it is a non-standard pattern.
+
+# Static Typing Concerns
+
+While these APIs will not be able to provide the same level of type safety as provided by GraphQL, we can try and make them easy 
+to add types to using generics. An important consideration is making it easy for the consumer to know what generics to pass. The
+best way of doing that is making them consistent across calls. For example, allowing the user to pass in a single type for all calls
+that matches the redux state, and the RPC handlers. Then we can use type helpers to get inference where needed. For example:
+
+```js
+// some-component.js
+import type {HandlersType} from './rpc/handlers';
+import useRPCRedux from 'fusion-plugin-rpc-redux-react';
+
+function component() {
+  const rpc = useRPCRedux<HandlersType>();
+  throw rpc.getUser({userId: 'abcd'});
+}
+
+// ...
+// rpc/handlers.js
+// ...
+type User = {
+  id: string,
+  name: string,
+};
+export type HandlersType = {
+  getUser(args: {userId: string}, ctx: Context): Promise<User>
+}
+
+// ... fusion-plugin-rpc-redux-react
+type ExtractArgumentType = <A>((A, any) => any) => (A) => any
+function useRPCRedux<A: Object>(): $ObjMap<A, ExtractArgumentType> {
+  // ... implementation ...
+}
+```
+
+This pattern allows the consumer to maintain a single type which represents the RPC handlers API, and share that type across their server
+and browser code.
+
+This gets a bit more difficult as the abstraction grows. Having this pattern apply to `createRPCResource` and `createRPCReactorResource` for example
+is much more difficult.
+
+# Drawbacks
+
+All the proposed API's are dependent on react suspense for data fetching. React suspense is still in an experimental phase and 
+there is no support for server side rendering. We can get support via packages such as react-ssr-prepass, but there is a risk
+that the behavior of this package will differ from the server side render behavior when react releases official support for suspense.
+
+It is not yet recommended to use react suspense for data fetching. The only officially supported usecase is React.lazy. There are 
+some examples of how to use the suspense pattern for data fetching in the react-cache package, but these patterns could change. There
+is a risk that new recommended patterns will emerge that don't match with our recommendations here. This risk increases as our level
+of abstraction increases. In other words, there is relatively low risk for `useRPC` and `useRPCRedux`, and high risk for `createRPCResource`
+and `createRPCReactorResource`.
```

Comment:
I agree - I think we should hold off on implementing the createResource like APIs

Reply:
Worried about the impact on flow typing, since it doesn't support typechecking thrown values: https://flow.org/try/#0GYVwdgxgLglg9mABAQwBQEpEG8BQBIKACwCc4B3RABVIFsYBnAUwDpjH64AbAN0dQHJ+6ANw4AvjijEAntnxoR4xBGRQIhRKkaYsiAPR7EAHgC0iTnHr1EcYIijSADo0QwwwOMRqr4SQozZ8FiJGMFQkAF4APkQwZig4ADEYAA9GABMMdHEgA

Reply:
You are correct, the errors will not be typed. However, this is handled in the error boundaries, not in the components directly.